source/common/common/utility.cc

```diff
@@ -248,12 +248,48 @@ std::string StringUtil::escape(const std::string& source) {
 }
 
 std::string AccessLogDateTimeFormatter::fromTime(const SystemTime& time) {
-  static DateFormatter default_date_format("%Y-%m-%dT%H:%M:%S");
+  static const char DefaultDateFormat[] = "%Y-%m-%dT%H:%M:%S.000Z";
+
+  struct CachedTime {
+    std::chrono::seconds epoch_time_seconds;
+    size_t formatted_time_length{0};
+    char formatted_time[32];
+  };
+  static thread_local CachedTime cached_time;
+
+  const std::chrono::milliseconds epoch_time_ms =
+      std::chrono::duration_cast<std::chrono::milliseconds>(time.time_since_epoch());
+
+  const std::chrono::seconds epoch_time_seconds =
+      std::chrono::duration_cast<std::chrono::seconds>(epoch_time_ms);
+
+  if (cached_time.formatted_time_length == 0 ||
+      cached_time.epoch_time_seconds != epoch_time_seconds) {
+    time_t time = static_cast<time_t>(epoch_time_seconds.count());
+    tm date_time;
+    gmtime_r(&time, &date_time);
+    cached_time.formatted_time_length =
+        strftime(cached_time.formatted_time, sizeof(cached_time.formatted_time), DefaultDateFormat,
+                 &date_time);
+    cached_time.epoch_time_seconds = epoch_time_seconds;
+  }
+
+  if (cached_time.formatted_time_length >= 4) {
+    // Overwrite the digits in the ".000Z" at the end of the string with the
+    // millisecond count from the input time.
+    size_t offset = cached_time.formatted_time_length - 4;
+    uint32_t msec = epoch_time_ms.count() % 1000;
+    cached_time.formatted_time[offset++] = ('0' + (msec / 100));
+    msec %= 100;
+    cached_time.formatted_time[offset++] = ('0' + (msec / 10));
+    msec %= 10;
+    cached_time.formatted_time[offset++] = ('0' + msec);
+  } else if (cached_time.formatted_time_length == 0) {
```

Comment:
Can this ever actually happen? Can we ASSERT or RELEASE_ASSERT this? If not, we will need to test this path probably via injecting OsSysCalls.